oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/RemoveAuthorizedClientReactiveOAuth2AuthorizationFailureHandler.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizedClientRepository;
+import org.springframework.security.oauth2.core.OAuth2AuthorizationException;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.util.Assert;
+import org.springframework.web.server.ServerWebExchange;
+import reactor.core.publisher.Mono;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An authorization failure handler that removes authorized clients from a
+ * {@link ServerOAuth2AuthorizedClientRepository}
+ * or a {@link ReactiveOAuth2AuthorizedClientService}.
+ * for specific OAuth 2.0 error codes.
+ *
+ * @author Phil Clay
+ * @since 5.3
+ */
+public class RemoveAuthorizedClientReactiveOAuth2AuthorizationFailureHandler implements ReactiveOAuth2AuthorizationFailureHandler {
+
+	/**
+	 * The default OAuth2 error codes that will trigger removal of the authorized client.
+	 * @see OAuth2ErrorCodes
+	 */
+	public static final Set<String>
```

Comment:
Hmm.  IMHO, I don't think the cached client should be removed for all failures.  That would be too aggressive, since not all OAuth2 error codes indicate that a token is invalid.

I'd prefer that the default implementation follows the spec as closely as possible, and only remove tokens for the OAuth2 error codes for which the spec says the token is invalid.

As a side note, these aren't HTTP statuses, they're OAuth2 error codes.  And they are the specific error codes that (according to the spec) indicate that a token is no longer valid, and a new token should be retrieved.  The only reason I included the ability to configure these error codes at all is because there might be some out-of-spec servers that return incorrect OAuth2 error codes for invalid tokens.  E.g. in https://github.com/spring-projects/spring-security/issues/7583#issuecomment-549086681, @janzyka mentioned that their corporate server returns "invalid_request" instead of "invalid_grant" when the token is no longer valid.

If more complex (out-of-spec) filtering is needed other than oauth2 error codes, then a user could create a new implementation of `ReactiveOAuth2AuthorizationFailureHandler`.

Reply:
Good points. We'll keep it.